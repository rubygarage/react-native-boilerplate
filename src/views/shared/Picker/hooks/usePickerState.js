import { useState } from 'react';
import * as R from 'ramda';

const usePickerState = (props) => {
  const {
    selectedValue,
    options,
    intl: { formatMessage },
    onValueChange,
  } = props;

  const [isOpen, setIsOpen] = useState(false);

  const detectLabel = () => {
    const option = R.find(R.propEq('value', selectedValue))(options);

    return R.prop('label', option) || formatMessage({ id: 'shared.na' });
  };

  const handlePickerToggle = () => {
    setIsOpen((prevIsOpen) => !prevIsOpen);
  };

  const handleValueChange = (value) => {
    onValueChange(value);

    handlePickerToggle();
  };

  return {
    isOpen,
    setIsOpen,
    detectLabel,
    handlePickerToggle,
    handleValueChange,
  };
};

export default usePickerState;
