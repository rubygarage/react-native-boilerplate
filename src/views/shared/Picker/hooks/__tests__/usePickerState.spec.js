import { renderHook, act } from '@testing-library/react-hooks';

import usePickerState from '../usePickerState';

describe('useInputState custom hook', () => {
  const props = {
    selectedValue: 'test',
    options: [
      { value: 'test', label: 'label' },
      { value: 'test1', label: 'label1' },
    ],
    intl: {
      formatMessage: jest.fn(),
    },
    onValueChange: jest.fn(),
  };

  it('matches snapshot', () => {
    const { result } = renderHook(() => usePickerState(props));

    expect(result.current).toMatchSnapshot();
  });

  describe('checks detectLabel method', () => {
    it('should return right label', () => {
      const { result } = renderHook(() => usePickerState(props));

      let label = null;

      act(() => {
        label = result.current.detectLabel();
      });

      expect(label).toBe('label');
    });

    it('should return right label', () => {
      const { result } = renderHook(() => usePickerState({
        ...props,
        selectedValue: 'test1',
      }));

      let label = null;

      act(() => {
        label = result.current.detectLabel();
      });

      expect(label).toBe('label1');
    });

    it('should return undefined', () => {
      const { result } = renderHook(() => usePickerState({
        ...props,
        selectedValue: 'some',
      }));

      let label = null;

      act(() => {
        label = result.current.detectLabel();
      });

      expect(label).toBeFalsy();
    });
  });

  it('checks handlePickerToggle method', () => {
    const { result } = renderHook(() => usePickerState(props));

    act(() => {
      result.current.handlePickerToggle();
    });

    expect(result.current.isOpen).toBe(true);

    act(() => {
      result.current.handlePickerToggle();
    });

    expect(result.current.isOpen).toBe(false);
  });

  it('checks handleValueChange method', () => {
    const value = 'test';

    const { result } = renderHook(() => usePickerState(props));

    act(() => {
      result.current.handleValueChange(value);
    });

    expect(result.current.isOpen).toBe(true);
    expect(props.onValueChange).toHaveBeenCalledWith(value);
  });
});
