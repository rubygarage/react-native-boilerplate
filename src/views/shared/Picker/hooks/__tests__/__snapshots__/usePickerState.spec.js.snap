// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useInputState custom hook matches snapshot 1`] = `
Object {
  "detectLabel": [Function],
  "handlePickerToggle": [Function],
  "handleValueChange": [Function],
  "isOpen": false,
  "setIsOpen": [Function],
}
`;
